tests/61push/08_rejected_pushers.pl

```diff
@@ -0,0 +1,104 @@
+sub create_pusher
+{
+   my ( $user, $app_id, $push_key, $url ) = @_;
+
+   do_request_json_for( $user,
+      method  => "POST",
+      uri     => "/r0/pushers/set",
+      content => {
+         profile_tag         => "tag1",
+         kind                => "http",
+         app_id              => $app_id,
+         app_display_name    => "sytest_display_name",
+         device_display_name => "device_display_name",
+         pushkey             => $push_key,
+         lang                => "en",
+         data                => { url => $url },
+      },
+   );
+}
+
+sub wait_for_push
+{
+   my ( $path, $response ) = @_;
+   await_http_request( $path, sub {
+     my ( $request ) = @_;
+     my $body = $request->body_from_json;
+
+     return unless $body->{notification}{type};
+     return unless $body->{notification}{type} eq "m.room.message";
+     return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+
+      $request->respond_json( $response // {} );
+      Future->done( $request );
+   });
+}
+
+multi_test "Test that rejected pushers are removed.",
+   requires => [
+      local_user_fixtures( 2, with_events => 0 ),
+      $main::TEST_SERVER_INFO,
+   ],
+
+   do => sub {
+      my ( $alice, $bob, $test_server_info ) = @_;
+
+      my $room_id;
+
+      my $url = $test_server_info->client_location . "/alice_push";
+
+      matrix_create_room( $alice, visibility => "private" )->then( sub {
+         ( $room_id ) = @_;
+         matrix_invite_user_to_room( $alice, $bob, $room_id ),
+      })->then( sub {
+         matrix_join_room( $bob, $room_id);
+      })->then( sub {
+         matrix_send_room_text_message(
+            $bob, $room_id, body => "message"
+         )->then( sub {
```

Comment:
Any reason this `->then` block is nested inside? It looks to me as if it can be out at the same level as the others